### PR TITLE
FIX theorical stock when stock calculate on bill

### PR DESCRIPTION
**Pull request by atm-quentin**
# FIX|Fix theorical stock when stock calculate on bill
When you have an order, you create an invoice and validate it the theorical stock is wrong on the product

### CODE REVIEW

File: htdocs/product/class/product.class.php [hunk 1]
```diff
@@ -2990,7 +2990,24 @@ public function load_stats_commande($socid = 0, $filtrestatut = '', $forVirtualS
...
 
+					$this->stats_commande['qty'] -= $adeduire;
+				} else {
+					//For every order having invoice already validated we need to decrease stock cause it's in physical stock
+					$adeduire = 0;
+					$sql = 'SELECT sum(fd.qty) as count FROM '.MAIN_DB_PREFIX.'facturedet fd ';
+					$sql .= ' JOIN '.MAIN_DB_PREFIX.'facture f ON fd.fk_facture = f.rowid ';
+					$sql .= ' JOIN '.MAIN_DB_PREFIX."element_element el ON el.fk_target = f.rowid and el.targettype = 'facture' and sourcetype = 'commande'";
+					$sql .= ' JOIN '.MAIN_DB_PREFIX.'commande c ON el.fk_source = c.rowid ';
+					$sql .= ' WHERE c.fk_statut IN ('.$this->db->sanitize($filtrestatut).') AND f.fk_statut > '.Facture::STATUS_DRAFT.' AND fd.fk_product = '.((int) $this->id);
```
Comment (eldy) [inline htdocs/product/class/product.class.php:3009]: If status is > Draft, it means invoice has been validated.
And we are in option STOCK_CALCULATE_ON_BILL is on meaning the real stock is changed when bill is validated.
If bill is validated (this first line), the real stock has been changed so there is no reason to include content of invoice in theorical stock.

File: htdocs/product/class/product.class.php [hunk 2]
```diff
@@ -2990,7 +2990,24 @@ public function load_stats_commande($socid = 0, $filtrestatut = '', $forVirtualS
...
+						}
+					}
 
+					$this->stats_commande['qty'] -= $adeduire;
+				} else {
+					//For every order having invoice already validated we need to decrease stock cause it's in physical stock
+					$adeduire = 0;
+					$sql = 'SELECT sum(fd.qty) as count FROM '.MAIN_DB_PREFIX.'facturedet fd ';
+					$sql .= ' JOIN '.MAIN_DB_PREFIX.'facture f ON fd.fk_facture = f.rowid ';
+					$sql .= ' JOIN '.MAIN_DB_PREFIX."element_element el ON el.fk_target = f.rowid and el.targettype = 'facture' and sourcetype = 'commande'";
```
Comment (eldy) [inline htdocs/product/class/product.class.php:3007]: link can be in 2 sides. So 
(el.fk_target = f.rowid and el.targettype = 'facture' and sourcetype = 'commande'")
must be complete with 
OR 
(el.fk_source = f.rowid and el.targettype = 'commande' and sourcetype = 'facture'")

**Comment by atm-quentin** [comment]
done

**Comment by atm-quentin** [comment]
@eldy

**Comment by atm-quentin** [comment]
@eldy

**Comment by atm-quentin** [comment]
@eldy

**Comment by eldy** [comment]
i added the 
if ($forVirtualStock
because method is used for stats on order (in such case, we must keep data brute from order with effect from invoices)
or be used for virtual stock calculation (in such a case, compensation for validated billed can be done)

**Pull request merged**